tokio/src/io/async_fd.rs

```diff
@@ -449,7 +540,28 @@ impl<T: AsRawFd> AsyncFd<T> {
         })
     }
 
-    async fn readiness_mut(
+    /// Waits for any of the requested ready states, returning a
+    /// [`AsyncFdReadyMutGuard`] that must be dropped to resume
+    /// polling for the requested ready states.
+    ///
+    /// The function may complete without the file descriptor being ready. This is a
+    /// false-positive and attempting an operation will return with
+    /// `io::ErrorKind::WouldBlock`. The function can also return with an empty
+    /// [`Ready`] set, so you should always check the returned value and possibly
+    /// wait again if the requested states are not set.
+    ///
+    /// When an IO operation does return `io::ErrorKind::WouldBlock`, the readiness must be cleared.
+    /// When a combined interest is used, it is important to clear only the readiness
+    /// that is actually observed to block. For instance when the combined
+    /// interest `Interest::READABLE | Interest::WRITABLE` is used, and a read blocks, only
+    /// read readiness should be cleared using the [`AsyncFdReadyMutGuard::clear_ready_exact`] method:
+    /// `guard.clear_ready_exact(Ready::READABLE)`.
+    /// Also clearing the write readiness in this case would be incorrect.
+    /// The [`AsyncFdReadyMutGuard::clear_ready`] method clears all readiness flags.
+    ///
+    /// This method takes `&mut self`, so it is possible to access the inner IO
+    /// resource mutably when handling the [`AsyncFdReadyMutGuard`].
+    pub async fn ready_mut(
```

Comment:
A doc example would be nice to have here. You could copy the one above and use some sort of mutation.

Reply:
based on some github searches, it looks like the `_mut` functions are hardly used at all. The only uses I could find are around linux eBPF stuff. Even there it might work to implement traits on a reference, e.g. there is a `impl Write for &UdpSocket` so that you don't need to have a mutable reference to a UdpSocket but can still implement the `Write` trait, which does take a mutable reference.

On the other hand, the function deserves an example, and I couldn't think of any better examples, so I've just replicated the one from `ready` here.

Reply:
A basic example is totally fine. Part of the reason is to just ensure we also have a test for the public API (via doc tests).